augur/tasks/util/collection_util.py

```diff
@@ -425,3 +433,155 @@ def start_block_of_repos(logger,session,condition,limit,phases,repos_type,hook="
+        JOIN augur_data.repo ON augur_operations.user_repos.repo_id = augur_data.repo.repo_id
+        JOIN augur_operations.collection_status ON augur_operations.user_repos.repo_id = augur_operations.collection_status.repo_id
+        WHERE {desired_status}='{str(new_status)}'
+        GROUP BY user_id
+    """)
+
+    user_list = session.execute_sql(query).fetchall()
+    random.shuffle(user_list)
+
+    #Extract the user id from the randomized list and split into four chunks
+    split_user_list = split_list_into_chunks([row[0] for row in user_list], 4)
+
+    session.logger.info(f"User_list: {split_user_list}")
+
+    #Iterate through each fourth of the users fetched
+    for quarter_list in split_user_list:
+        if limit <= 0:
+            return
+
+        #Query a set of valid repositories sorted by weight, also making sure that the repos are new
+        repo_query = s.sql.text(f"""
+            SELECT
+	        repo.repo_id
+            FROM augur_operations.user_groups 
+            JOIN augur_operations.user_repos ON augur_operations.user_groups.group_id = augur_operations.user_repos.group_id
+            JOIN augur_data.repo ON augur_operations.user_repos.repo_id = augur_data.repo.repo_id
+            JOIN augur_operations.collection_status ON augur_operations.user_repos.repo_id = augur_operations.collection_status.repo_id
+            WHERE user_id IN :list_of_user_ids AND {desired_status}='{str(new_status)}'
+            ORDER BY {hook}_weight
+            LIMIT :limit_num
+        """).bindparams(list_of_user_ids=tuple(quarter_list),limit_num=(limit*2))
+
+        #Get a list of valid repo ids, limit set to 2 times the usual
+        valid_repos = session.execute_sql(repo_query).fetchall()
+        valid_repo_git_ids = [repo[0] for repo in valid_repos]
+
+        #Create conditions for this query of valid repos to make sure they are new and valid
+        session.logger.info(f"valid repo git ids: {tuple(valid_repo_git_ids)}")
+        never_collected = getattr(CollectionStatus,f"{hook}_data_last_collected" ) == None
+        make_sure_valid_repo = CollectionStatus.repo_id.in_(list(valid_repo_git_ids))
+
+        #Create a list of conditions and add any more if passed in
+        cond_list = [not_erroed, not_collecting,never_collected,make_sure_valid_repo]
+
+        if additional_conditions:
+            cond_list.extend(additional_conditions)
+
+        condition = and_(*cond_list)
+
+        #Order by the relevant weight for the collection hook
+        order = getattr(CollectionStatus,f"{hook}_weight" )
+
+        #start repos for new primary collection hook
+        collection_size = start_block_of_repos(
+            session.logger, session,
+            condition,
+            limit, phase_list, repos_type="new", hook=hook,sort=order
+        )
+        
+        #Update limit with amount of repos started
+        limit -= collection_size
+
+    #Now start old repos if there is space to do so.
```

Comment:
I believe there should still be a check here to see if the the limit is less than or equal to 0. To cover a case like this: The loop is in its last iteration and the limit is 5. Then 5 repos are started so the limit is 0. The loop then exits, so the limit check in the for loop is never reached to return. So it would continue on to start new repos even though the limit is 0

Reply:
It would return before it would start any repos. It would query the users table unnecessarily though

Reply:
Yeah, I didn't mean that it would start any repos, but I think we should add this check because it removes an unnecessary query like you mentioned and makes the code easier to read since it is easy to see how the code fulfills this comment "Now start old repos if there is space to do so."

Reply:
That makes sense